### ISSUE TITLE
bug: Equivalent code works in tensorflow python but not tensorflow-js

### ISSUE THREAD

**Issue by shepherdj2** [tensorflow/tfjs#1773]
#### TensorFlow.js version
Tried in versions `1.1.2` and `1.2.5`

#### Describe the problem or feature request

The same network works in python version of tensorflow but the equivalent code does not work with tensorflow-js resulting in the error `Error: Size(9) must match the product of shape 1,1,1,1`

#### Code to reproduce the bug / link to feature request

**tensorflow-js version**

```Javascript
const tf = require('@tensorflow/tfjs');
require('@tensorflow/tfjs-node');  // Use '@tensorflow/tfjs-node-gpu' if running with GPU.
const { regularizers } = tf;
const { conv2d, batchNormalization, leakyReLU, input, dense, flatten } = tf.layers;

let main_input = input({ shape: [2, 3, 3], name: 'main_input' })

let x = conv2d({
  kernelSize: [4, 4],
  filters: 75,
  activation: 'linear',
  dataFormat: "channelsFirst",
  padding: 'same',
  useBias: false,
  kernelRegularizer: regularizers.l2(0.0001)
}).apply(main_input)

x = batchNormalization({ axis: 1 }).apply(x)
x = leakyReLU().apply(x)

x = conv2d({
  filters: 1,
  kernelSize: [1, 1],
  activation: 'linear',
  dataFormat: "channelsFirst",
  padding: 'same',
  name: 'value_head_conv2d',
  useBias: false,
  kernelRegularizer: regularizers.l2(0.0001)
}).apply(x)

x = batchNormalization({ axis: 1, name: 'value_head_batch_normalization' }).apply(x)
x = leakyReLU({ name: 'value_head_leaky_relu_1' }).apply(x)

x = flatten({ name: 'value_head_flatten' }).apply(x)

x = dense({
  units: 1,
  useBias: false,
  activation: 'tanh',
  kernelRegularizer: regularizers.l2(0.0001),
  name: 'output_1'
}).apply(x)

let model = tf.model({ inputs: [main_input], outputs: [x] })
model.compile({
  loss: { 'output_1': 'meanSquaredError' },
  optimizer: tf.train.sgd(0.1),
})

let model_input = tf.tensor([[[[1,0,0],[1,1,0],[0,0,0]],[[0,0,1],[0,0,0],[1,1,0]]]])
let model_output = tf.tensor([[1]])

model.fit(model_input, { 'output_1': model_output }, {
  epochs: 10,
  callbacks: {
    onEpochEnd: async (epoch, log) => {
      console.log(`Epoch ${epoch}: loss = ${log.loss}`);
    }
  }})
```

**python version**

```Python
import tensorflow as tf

from keras.models import Sequential, load_model, Model
from keras.layers import Input, Dense, Conv2D, Flatten, BatchNormalization, Activation, LeakyReLU, add
from keras.optimizers import SGD
from keras import regularizers

import numpy as np

main_input = Input(shape = [2, 3, 3], name = 'main_input')

x = Conv2D(
		filters = 75
		, kernel_size = (4,4)
		, data_format="channels_first"
		, padding = 'same'
		, use_bias=False
		, activation='linear'
		, kernel_regularizer = regularizers.l2(0.0001)
		)(main_input)

x = BatchNormalization(axis=1)(x)
x = LeakyReLU()(x)

x = Conv2D(
		filters = 1
		, kernel_size = (1,1)
		, data_format="channels_first"
		, padding = 'same'
		, use_bias=False
		, activation='linear'
		, kernel_regularizer = regularizers.l2(0.0001)
		)(x)


x = BatchNormalization(axis=1)(x)
x = LeakyReLU()(x)

x = Flatten()(x)

x = Dense(
  1
  , use_bias=False
  , activation='tanh'
  , kernel_regularizer=regularizers.l2(0.0001)
  , name = 'output_1'
  )(x)

model = Model(inputs=[main_input], outputs=[x])

model.compile(loss={'output_1': 'mean_squared_error'},
			optimizer=SGD(lr=0.001, momentum = 0.001)	
			)

model_input = np.array([[[[1,0,0],[1,1,0],[0,0,0]],[[0,0,1],[0,0,0],[1,1,0]]]])

model_output = np.array([[1]])

model.fit(model_input, {'output_1': model_output}, epochs= 10)
```

### results

**tensorflowjs result:** ` Error: Size(9) must match the product of shape 1,1,1,1`
**python result:** runs training as expected, no errors.

**Comment by rthadur** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow.js) since it is not a bug or feature request. There is also a larger community that reads questions there.

**Comment by shepherdj2** [comment]
@rthadur I open the issue here and not on StackOverflow because this looks like a tensorflow-js BUG and the comparison to the python version is meant to exemplify that.

**Comment by rthadur** [comment]
no worries , reopening it , @annxingyuan can you please help with this issue ?

**Comment by waveboo** [comment]
I have encountered the same problem. Anybody solved this problem?
And I found that when I load the pre-trained model of keras using tensorflow,.js the model can predict the input data, but still can't be trained.

**Comment by caisq** [comment]
@waveboo @shepherdj2 I'm looking into this issue now. So I noticed two things.

1. This is not directly related to the error you are getting, but should be fixed as well. Specifically, when you call the L2 regularizer in JavaScript, you should make sure you use `l2({l2: 0.0001)` instead of `l2(0.0001)`. The latter doesn't do what you want and it currently doesn't throw an error. We'll fix that so that it errors out.

2. You are using `axis: 1` for your BatchNormalization layers. Is that because you are using channelFirst? The axis: 1 is what's directly cause the error " Error: Size(9) must match the product of shape 1,1,1,1". Using the default `axis: -1` can get you around the issue for now. So to unblock yourself, you may switch to channelLast, which will involve changes to your conv2d code and your input data as well. We'll fix the `axis: 1` issue in the meantime.

**Comment by rthadur** [comment]
@shepherdj2 @waveboo there is fix which is merged now , can you please verify once ?

**Comment by caisq** [comment]
@rthadur The PR that just got merge (#1861) fixes issue 1 mentioned above. Issue #2, which is more fundamental, hasn't been addressed yet, which is why this issue should be left open.

**Comment by rthadur** [comment]
@caisq thank you will keep this open.

**Comment by lp74** [comment]
I've got the same problem.
I am just building a simple regularized (l2) linear regression:


`const model = tf.sequential();
model.add(tf.layers.dense({
    units: 1,
    inputShape: [1], 
    useBias: true,
    kernelRegularizer: tf.regularizers.l2(0.01),
}));
`
It seems that the regularizer doesn't work.

Any hint?

**Comment by caisq** [comment]
@lp74 Please construct the regularizer with an object argument: `{l2: 0.01}`, instad of `0.01`.

**Comment by gaikwadrahul8** [comment]
Hi, @shepherdj2 

Apologize for the delayed response and I see above both PR's https://github.com/tensorflow/tfjs/pull/1861 and https://github.com/tensorflow/tfjs/pull/2 got merged to take care of this issue so if issue got resolved could you please close this issue now ? Thank you!
